### PR TITLE
Disconnect announcement_signatures sending from funding_locked

### PR DESCRIPTION
**Pull request by TheBlueMatt**
The spec actually requires we never send `announcement_signatures`
(and, thus, `channel_announcement`s) until after six confirmations.
However, we would happily have sent them prior to that as long as
we exchange `funding_locked` messages with our countarparty. Thanks
to re-broadcasting this issue is largely harmless, however it could
have some negative interactions with less-robust peers. Much more
importantly, this represents an important step towards supporting
0-conf channels, where `funding_locked` messages may be exchanged
before we even have an SCID to construct the messages with.

Because there is no ACK mechanism for `announcement_signatures` we
rely on existing channel updates to stop rebroadcasting them - if
we sent a `commitment_signed` after an `announcement_signatures`
and later receive a `revoke_and_ack`, we know our counterparty also
received our `announcement_signatures`. This may resolve some rare
edge-cases where we send a `funding_locked` which our counterparty
receives, but lose connection before the `announcement_signatures`
(usually the very next message) arrives.

Sadly, because the set of places where an `announcement_signatures`
may now be generated more closely mirrors where `funding_locked`
messages may be generated, but they are now separate, there is a
substantial amount of code motion providing relevant parameters
about current block information and ensuring we can return new
`announcement_signatures` messages.

### CODE REVIEW

File: lightning/src/ln/channelmanager.rs [hunk 1]
```diff
@@ -4872,23 +4861,34 @@ where
...
 						short_to_id.insert(channel.get_short_channel_id().unwrap(), channel.channel_id());
 					}
+					if let Some(announcement_sigs) = announcement_sigs {
+						log_trace!(self.logger, "Sending announcement_signatures for channel {}", log_bytes!(channel.channel_id()));
+						pending_msg_events.push(events::MessageSendEvent::SendAnnouncementSignatures {
+							node_id: channel.get_counterparty_node_id(),
+							msg: announcement_sigs,
+						});
+						if let Some(height) = height_opt {
+							if let Some(announcement) = channel.get_signed_channel_announcement(self.get_our_node_id(), self.genesis_hash, height) {
```
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs:4882]: This seems to be a behavior change that could be worth documenting in the commit msg

File: lightning/src/ln/functional_tests.rs [hunk 2]
```diff
@@ -426,10 +426,52 @@ fn do_test_1_conf_open(connect_style: ConnectStyle) {
...
+	if let MessageSendEvent::SendChannelUpdate { ref node_id, msg: _ } = bs_msg_events[0] {
+		assert_eq!(*node_id, nodes[0].node.get_our_node_id());
+	} else { panic!("Unexpected event"); }
+
+	send_payment(&nodes[0], &[&nodes[1]], 100_000);
+
+	// After 6 confirmations, as required by the spec, we'll send announcement_signatures and
+	// broadcast the channel_announcement.
+	connect_blocks(&nodes[0], 5);
+	nodes[1].node.handle_announcement_signatures(&nodes[0].node.get_our_node_id(), &get_event_msg!(nodes[0], MessageSendEvent::SendAnnouncementSignatures, nodes[1].node.get_our_node_id()));
```
Comment (valentinewallace) [inline lightning/src/ln/functional_tests.rs:454]: Could check it isn't generated a block early

File: lightning/src/ln/channel.rs [hunk 3]
```diff
@@ -3203,6 +3249,8 @@ impl<Signer: Sign> Channel<Signer> {
 			})
 		} else { None };
 
+		let announcement_sigs = self.get_announcement_sigs(node_pk, genesis_block_hash, best_block_height).ok();
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:3252]: `get_announcement_sigs` calls out to `get_channel_announcement` which returns an err if `!self.config.announced_channel`.

File: lightning/src/ln/channelmanager.rs [hunk 4]
```diff
@@ -3604,8 +3596,8 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 				return;
 			}
 
-			let updates = channel.get_mut().monitor_updating_restored(&self.logger);
-			let channel_update = if updates.funding_locked.is_some() && channel.get().is_usable() && !channel.get().should_announce() {
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3621]: Yep, good point. Updated the comment. There's not really a good reason for or against sending it twice here, its kinda nice to just send it and remove the check to simplify things down, though, IMO.

File: lightning/src/ln/channelmanager.rs (hunk 1 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:4882]: Wait, which part? The commit message explicitly states
```
    The spec actually requires we never send `announcement_signatures`
    (and, thus, `channel_announcement`s) until after six confirmations.
    However, we would happily have sent them prior to that as long as
    we exchange `funding_locked` messages with our countarparty.
```

**Review by valentinewallace — COMMENTED** [review]
Shaping up!

File: lightning/src/ln/channel.rs [hunk 5]
```diff
@@ -4229,8 +4349,11 @@ impl<Signer: Sign> Channel<Signer> {
 		self.update_time_counter = cmp::max(self.update_time_counter, highest_header_time);
 
 		if let Some(funding_locked) = self.check_get_funding_locked(height) {
+			let announcement_sigs = if let Some((genesis_block_hash, node_pk)) = genesis_node_pk {
+				self.get_announcement_sigs(node_pk, genesis_block_hash, height, logger)
+			} else { None };
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:4354]: `get_announcement_sigs` updates our "message sent" state, so we really can't call it unless we're definitely about to return `Ok`.

File: lightning/src/ln/channel.rs [hunk 6]
```diff
@@ -4182,7 +4296,8 @@ impl<Signer: Sign> Channel<Signer> {
 					// may have already happened for this block).
 					if let Some(funding_locked) = self.check_get_funding_locked(height) {
 						log_info!(logger, "Sending a funding_locked to our peer for channel {}", log_bytes!(self.channel_id));
-						return Ok(Some(funding_locked));
+						let announcement_sigs = self.get_announcement_sigs(node_pk, genesis_block_hash, height, logger);
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:4299]: `check_get_funding_locked` updates our local state flags, though, so it has to come after that.

File: lightning/src/ln/channel.rs [hunk 7]
```diff
@@ -3192,6 +3252,8 @@ impl<Signer: Sign> Channel<Signer> {
 			})
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:3252]: `check_get_funding_locked` is a bit of a misnomer - it updates state and isn't really just a getter, its really just part of the block/transaction-connection logic refactored out into a poorly-named function :/

**Comment by TheBlueMatt** [comment]
Squashed.

**Comment by TheBlueMatt** [comment]
Rebased to update the TLV numbering due to conflicts.

**Review by arik-so — APPROVED** [review]

**Review by arik-so — APPROVED** [review]
re-ACKing

File: lightning/src/ln/channel.rs [hunk 8]
```diff
@@ -4596,25 +4723,24 @@ impl<Signer: Sign> Channel<Signer> {
 		})
 	}
 
-	/// Gets an UnsignedChannelAnnouncement, as well as a signature covering it using our
-	/// bitcoin_key, if available, for this channel. The channel must be publicly announceable and
-	/// available for use (have exchanged FundingLocked messages in both directions). Should be used
-	/// for both loose and in response to an AnnouncementSignatures message from the remote peer.
+	/// Gets an UnsignedChannelAnnouncement for this channel. The channel must be publicly
+	/// announceable and available for use (have exchanged FundingLocked messages in both
+	/// directions). Should be used for both loose and in response to an AnnouncementSignatures
```
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/ln/channel.rs]: Was this resolved...?

[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs]: I reworked the documentation to no longer say "loose".

**Comment by arik-so** [comment]
One aside: it's "Disconnect" with two _n_s.

File: lightning/src/ln/channel.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/ln/channel.rs]: Ah, it looks like the updated comment was put in ed1163a5bfa3ce02daaee505df33fef119556b25

**Review by valentinewallace — APPROVED** [review]
Some nits. ACK ed1163a5bfa3ce02daaee505df33fef119556b25

File: lightning/src/ln/channel.rs [hunk 9]
```diff
@@ -4596,17 +4596,19 @@ impl<Signer: Sign> Channel<Signer> {
...
+	/// message from the remote peer.
+	///
 	/// Will only fail if we're not in a state where channel_announcement may be sent (including
 	/// closing).
+	///
 	/// Note that the "channel must be funded" requirement is stricter than BOLT 7 requires - see
 	/// https://github.com/lightningnetwork/lightning-rfc/issues/468
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:4735]: Will fix in a followup PR, since this has two acks.

**Pull request merged**